Pod/Classes/VOKCoreDataManager.m

```diff
@@ -484,78 +495,100 @@ - (NSManagedObjectContext *)safeContext:(NSManagedObjectContext *)context
-        }];
-    }
+    [self saveContext:self.managedObjectContext andWait:NO];
 }
 
 - (void)saveMainContextAndWait
 {
-    if ([NSOperationQueue mainQueue] == [NSOperationQueue currentQueue]) {
-        [self saveContext:self.managedObjectContext];
-    } else {
-        [self.managedObjectContext performBlockAndWait:^{
-            [self saveContext:self.managedObjectContext];
-        }];
-    }
+    [self saveContext:self.managedObjectContext andWait:YES];
 }
 
-- (void)saveContext:(NSManagedObjectContext *)context
+- (void)obtainPermanentIDsForInsertionsInContext:(NSManagedObjectContext *)context
 {
-    NSError *error;
-    if (![context save:&error]) {
-        VOK_CDLog(@"Unresolved error %@, %@", error, [error localizedDescription]);
-    }
+    [context performBlockAndWait:^{
+        NSError *error;
+        if (![context obtainPermanentIDsForObjects:context.insertedObjects.allObjects
+                                             error:&error]) {
+            VOK_CDLog(@"Error obtaining permanent ID for insertions: %@", error);
+        }
+    }];
 }
 
-- (void)saveTempContext:(NSManagedObjectContext *)tempContext
-{
-    [[NSNotificationCenter defaultCenter] addObserver:self
-                                             selector:@selector(tempContextSaved:)
-                                                 name:NSManagedObjectContextDidSaveNotification
-                                               object:tempContext];
+/**
+ Save any changes to a managed object context to the persistent store.
+ If a context has no changes, no saves will be performed.
+ 
+ @param context The context to save
+ @param wait Whether to save synchronously (YES) or asynchronously (NO)
+ */
+- (void)saveContext:(NSManagedObjectContext *)context
+            andWait:(BOOL)wait
+{
+    //The documentation indicates that this method is not thread safe.
+    //We can't guarantee that the current method is called on the appropriate thread for this context,
+    //so use performBlockAndWait to query the context for changes.
+    __block BOOL hasChanges;
+    [context performBlockAndWait:^{
+        hasChanges = context.hasChanges;
+    }];
     
-    [self saveContext:tempContext];
+    if (!hasChanges) {
+        return;
+    }
     
-    [[NSNotificationCenter defaultCenter] removeObserver:self
-                                                    name:NSManagedObjectContextDidSaveNotification
-                                                  object:tempContext];
-}
-
-- (void)tempContextSaved:(NSNotification *)notification
-{
-    // Solved issue with NSFetchedResultsController ignoring changes
-    // merged from different managed object contexts by touching
-    // willAccessValueForKey: on the updated objects.
-    //http://stackoverflow.com/questions/3923826/nsfetchedresultscontroller-with-predicate-ignores-changes-merged-from-different
+    [self obtainPermanentIDsForInsertionsInContext:context];
```

Comment:
Why is it necessary to `obtainPermanentIDsForInsertionsInContext:` before saving? (Sorry if I missed an earlier discussion)